File: WebClient SDK/src/default/forums/extensions/krSettingsForumsSettingsUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { GridRowAction, GridViewModel } from 'tessa/ui/cards/controls';
import { ForumHelper } from 'tessa/forums';
import { WfUiHelper } from '../../workflow/wf/wfUiHelper';
import { extension } from '@tessa/application';
import { DisposeList, StorageHelper, TypedField } from '@tessa/core';
import { CardRow, CardRowsListener, WorkflowCardTypes } from '@tessa/platform';
import { Visibility } from 'tessa/platform';

@extension()
export class KrSettingsForumsSettingsUIExtension extends CardUIExtension {
  cardRowsListener: CardRowsListener | null;
  readonly disposeList = new DisposeList();

  public initialized(context: ICardUIExtensionContext): void {
    const { model } = context;
    if (model.inSpecialMode) {
      return;
    }

    const hasForumLicense =
      StorageHelper.tryGet<boolean>(model.card.tryGetInfo(), ForumHelper.LicenseWarningFlag) !==
      true;

    const cardTypeId = model.card.typeId;
    if (cardTypeId === WorkflowCardTypes.KrSettingsTypeID) {
      const sections = model.card.tryGetSections();

      const cardTypesSection = sections?.tryGet('KrSettingsCardTypes');
      const cardTypesControl = model.controls.get('CardTypeControl');

      if (cardTypesSection && cardTypesControl) {
        const cardTypesGrid = cardTypesControl as GridViewModel;

        cardTypesGrid.rowInvoked.add(e => {
          if (e.action === GridRowAction.Inserted || e.action === GridRowAction.Opening) {
            const forumsBlock = e.rowModel?.blocks.get('UseForumBlock');

            if (forumsBlock) {
              const useForumFirst = e.row.get<boolean>(ForumHelper.UseForumField) ?? false;

              WfUiHelper.setControlVisibility(
                forumsBlock,
                ForumHelper.UseForumSuffix,
                useForumFirst
              );

              const dispose = e.row.fieldChanged.add((e, s) => {
                if (e.fieldName === ForumHelper.UseForumField) {
                  const useForum = e.fieldValue as boolean;
                  s.set(
                    ForumHelper.UseDefaultDiscussionTabField,
                    TypedField.createBoolean(useForum)
                  );
                  WfUiHelper.setControlVisibility(
                    forumsBlock,
                    ForumHelper.UseForumSuffix,
                    useForum
                  );
                }
              });

              if (dispose) {
                forumsBlock.form.closed.addOnce(() => {
                  dispose();
                });
              }
            }

            if (!hasForumLicense) {
              const warningLabel = e.rowModel?.controls.get(ForumHelper.LicenseWarningControlAlias);
              if (warningLabel) {
                warningLabel.controlVisibility = Visibility.Visible;
              }
            }
          }
        });

        for (const row of cardTypesSection.rows) {
          this.attachHandlersToCardTypeRow(row);
        }

        this.cardRowsListener = new CardRowsListener();
        this.cardRowsListener.rowInserted.add(({ row }) => this.attachHandlersToCardTypeRow(row));
        this.cardRowsListener.start(cardTypesSection.rows);
      }
    } else if (cardTypeId === WorkflowCardTypes.KrDocTypeTypeID) {
      const sections = model.card.tryGetSections();
      if (sections) {
        const docTypeSection = sections.tryGet('KrDocType');
        const useForumBlock = model.blocks.get('UseForumBlock');
        if (docTypeSection && useForumBlock) {
          const useForumFirst =
            docTypeSection.fields.get<boolean>(ForumHelper.UseForumField) ?? false;
          WfUiHelper.setControlVisibility(useForumBlock, ForumHelper.UseForumSuffix, useForumFirst);

          const dispose = docTypeSection.fields.fieldChanged.add((e, s) => {
            if (e.fieldName === ForumHelper.UseForumField) {
              const useForum = e.fieldValue as boolean;
              s.set(ForumHelper.UseDefaultDiscussionTabField, TypedField.createBoolean(useForum));
              WfUiHelper.setControlVisibility(useForumBlock, ForumHelper.UseForumSuffix, useForum);
            }
          });

          if (dispose) {
            this.disposeList.add(dispose);
          }
        }
      }

      if (!hasForumLicense) {
        const warningLabel = model.controls.get(ForumHelper.LicenseWarningControlAlias);
        if (warningLabel) {
          warningLabel.controlVisibility = Visibility.Visible;
        }
      }
    }
  }

  public finalized(): void {
    if (this.cardRowsListener) {
      this.cardRowsListener.stop();
      this.cardRowsListener = null;
    }

    this.disposeList.dispose();
  }

  private attachHandlersToCardTypeRow(cardTypeRow: CardRow): void {
    const dispose = cardTypeRow.fieldChanged.add((e, s) => {
      if (e.fieldName === ForumHelper.UseForumField && !(e.fieldValue as boolean)) {
        s.set(
          ForumHelper.UseDefaultDiscussionTabField,
          TypedField.createBoolean(e.fieldValue as boolean)
        );
      }
    });

    if (dispose) {
      this.disposeList.add(dispose);
    }
  }
}

File: Extensions/Tessa.Extensions.Default.Client/Forums/KrSettingsForumsSettingsUIExtension.cs
#nullable enable
using System;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Default.Client.UI;
using Tessa.Extensions.Default.Client.Workflow.Wf;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Default.Shared.Workflow.Wf;
using Tessa.Forums;
using Tessa.Platform;
using Tessa.Platform.Storage;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Default.Client.Forums
{
    /// <summary>
    /// Визуальное расширение для карточки настроек типов карточек и типа документов,
    /// связанных с системой форумов.
    /// </summary>
    public sealed class KrSettingsForumsSettingsUIExtension : CardUIExtension
    {
        #region Base Overrides

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            var model = context.Model;
            if (model.InSpecialMode())
            {
                return;
            }

            var hasForumLicense = context.Card.TryGetInfo()?.TryGet<bool>(ForumHelper.LicenseWarningFlag) != true;

            var cardTypeID = model.CardType.ID;
            if (cardTypeID == DefaultCardTypes.KrSettingsTypeID)
            {
                if (context.Card.TryGetSections()?.TryGet(WfHelper.CardTypeSettingsSection) is { } section
                    && model.Controls.TryGet<GridViewModel>(KrTypesUIHelper.TypesControl) is { } grid)
                {
                    // скрываем / отображаем контролы, связанные с использованием резолюции
                    grid.RowInvoked += (s, e) =>
                    {
                        if (e.Action is GridRowAction.Inserted or GridRowAction.Opening)
                        {
                            if (e.RowModel.Blocks.TryGet(KrTypesUIHelper.UseForumBlock) is { } forumsBlock)
                            {
                                var useForumFirst = e.Row.Get<bool>(ForumHelper.UseForumField);
                                WfUIHelper.SetControlVisibility(forumsBlock, ForumHelper.UseForumSuffix, useForumFirst);
                                EventHandler<CardFieldChangedEventArgs> fieldChangedHandler = (s2, e2) =>
                                {
                                    if (e2.FieldName == ForumHelper.UseForumField)
                                    {
                                        var useForum = (bool) NotNullOrThrow(e2.FieldValue);

                                        ((ICardFieldContainer) NotNullOrThrow(s2)).Fields[ForumHelper.UseDefaultDiscussionTabField] = BooleanBoxes.Box(useForum);
                                        WfUIHelper.SetControlVisibility(forumsBlock, ForumHelper.UseForumSuffix, useForum);
                                    }
                                };

                                var row = e.Row;
                                row.FieldChanged += fieldChangedHandler;
                                forumsBlock.Form.Closed += (_, _) => row.FieldChanged -= fieldChangedHandler;
                            }

                            if (!hasForumLicense
                                && e.RowModel.Controls.TryGet(ForumHelper.LicenseWarningControlAlias) is { } warningLabel)
                            {
                                warningLabel.ControlVisibility = Visibility.Visible;
                                e.RowModel.MainFormWithBlocks.RearrangeSelf();
                            }
                        }
                    };

                    // сбрасываем состояние полей, когда галка "использовать систему форумов" очищается
                    var cardTypeRows = section.Rows;
                    foreach (var row in cardTypeRows)
                    {
                        row.FieldChanged += OnRowFieldChanged;
                    }

                    cardTypeRows.ItemChanged += static (s, e) =>
                    {
                        // чтобы не было утечек памяти при удалении строки внутри обработчика нельзя использовать замыканий
                        if (e is { Action: ListStorageAction.Insert, Item: { } item })
                        {
                            item.FieldChanged += OnRowFieldChanged;
                        }
                    };

                    static void OnRowFieldChanged(object? s, CardFieldChangedEventArgs e)
                    {
                        if (e.FieldName == ForumHelper.UseForumField)
                        {
                            var useForum = (bool) NotNullOrThrow(e.FieldValue);
                            ((ICardFieldContainer) NotNullOrThrow(s)).Fields[ForumHelper.UseDefaultDiscussionTabField] = BooleanBoxes.Box(useForum);
                        }
                    }
                }
            }
            else if (cardTypeID == DefaultCardTypes.KrDocTypeTypeID)
            {
                if (context.Card.TryGetSections()?.TryGet(WfHelper.DocTypeSettingsSection) is { } section
                    && model.Blocks.TryGet(KrTypesUIHelper.UseForumBlock, out var useForumBlock))
                {
                    // скрываем / отображаем контролы, связанные с использованием резолюции
                    var useForumFirst = section.RawFields.Get<bool>(ForumHelper.UseForumField);
                    WfUIHelper.SetControlVisibility(useForumBlock, ForumHelper.UseForumSuffix, useForumFirst);

                    section.FieldChanged += (s, e) =>
                    {
                        if (e.FieldName == ForumHelper.UseForumField)
                        {
                            // сбрасываем состояние полей, когда галка "использовать систему форумов" очищается
                            var useForum = (bool) NotNullOrThrow(e.FieldValue);
                            ((ICardFieldContainer) NotNullOrThrow(s)).Fields[ForumHelper.UseDefaultDiscussionTabField] = BooleanBoxes.Box(useForum);

                            // скрываем / отображаем контролы, связанные с использованием резолюции
                            WfUIHelper.SetControlVisibility(useForumBlock, ForumHelper.UseForumSuffix, useForum);
                        }
                    };
                }
                
                if (!hasForumLicense
                    && context.Model.Controls.TryGet(ForumHelper.LicenseWarningControlAlias) is { } warningLabel)
                {
                    warningLabel.ControlVisibility = Visibility.Visible;
                    warningLabel.Block.Form.RearrangeSelf();
                }
            }
        }

        #endregion
    }
}
